src/api/nbt_template.py

```diff
@@ -0,0 +1,333 @@
+from __future__ import annotations
+
+from typing import Union
+from typing import Any, Dict
+import json
+from collections import UserDict, UserList
+from glob import iglob
+from itertools import chain
+from os.path import join, basename
+import copy
+
+
+class NBTStruct:
+    """
+    Base class implemented by all NBT structure classes
+    """
+
+    def apply_to(self, other_tag: Union[NBTEntry, NBTListEntry, NBTCompoundEntry]):
```

Comment:
I went back and forth on this for awhile. I felt that it would make the most sense that the template would be "in charge" of dictating/running the templating process, so I implemented as "struct.apply(entry)". As for returning an entry, the return value doesn't really matter, since the top-level `NBTCompoundEntry` is eventually mutated, it's just returned to fit more in line with the `apply_to()` functions of `NBTStructure`/`NBTListStructure`

Reply:
Oh, wait. Is it mutating the `other_tag` parameter? That is a bit weird and unintuitive in my opinion. If it was a function inside `NBTCompoundEntry` rather than in the template, that would make more sense but since it's a parameter the function is getting, it's a bit weird.
I still think making it part of the Entry makes more sense. Can you explain why you ended up choosing putting it here?

Reply:
After thinking it over a bit, I think you're right about moving it, but that will depend on how we approach https://github.com/Amulet-Team/Amulet-Map-Editor/pull/36#discussion_r250400631